In [1]:
 %load_ext sql


In [2]:
%sql postgresql://postgres:0853@localhost/retail_revenue

## 1. Business Objective 

A growing online retailer of Sport/Athleisure products requests an exploratory analysis of their revenue stream to identify opportunities of improving revenue.

## 2. Description of dataset
The database provided to me, *Sports*, contains five tables, with product_id being the primary key for all of them:
<h3 id="info"><code>info</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>description</code></td>
<td><code>varchar</code></td>
<td>Description of the product</td>
</tr>
</tbody>
</table>
<h3 id="finance"><code>finance</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>listing_price</code></td>
<td><code>float</code></td>
<td>Listing price for product</td>
</tr>
<tr>
<td><code>sale_price</code></td>
<td><code>float</code></td>
<td>Price of the product when on sale</td>
</tr>
<tr>
<td><code>discount</code></td>
<td><code>float</code></td>
<td>Discount, as a decimal, applied to the sale price</td>
</tr>
<tr>
<td><code>revenue</code></td>
<td><code>float</code></td>
<td>Amount of revenue generated by each product, in US dollars</td>
</tr>
</tbody>
</table>
<h3 id="reviews"><code>reviews</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>rating</code></td>
<td><code>float</code></td>
<td>Product rating, scored from <code>1.0</code> to <code>5.0</code></td>
</tr>
<tr>
<td><code>reviews</code></td>
<td><code>float</code></td>
<td>Number of reviews for the product</td>
</tr>
</tbody>
</table>
<h3 id="traffic"><code>traffic</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>last_visited</code></td>
<td><code>timestamp</code></td>
<td>Date and time the product was last viewed on the website</td>
</tr>
</tbody>
</table>
<h3 id="brands"><code>brands</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>brand</code></td>
<td><code>varchar</code></td>
<td>Brand of the product</td>
</tr>
</tbody>
</table>

 ## 3. Checking data quality 

The dataset contains numeric, string, and timestamp data types. I start by checking how complete the data is. 

In [3]:
%%sql

-- Count all columns as total_rows
-- Count the number of non-missing entries for description, listing_price, and last_visited
-- Join info, finance, and traffic

SELECT COUNT(*) AS total_rows,
    COUNT(i.description) AS count_description,
    COUNT(f.listing_price) AS count_listing_price,
    COUNT (f.sale_price) AS count_sale_price,
    COUNT (f.discount) AS count_discount,
    COUNT (f.revenue) AS count_revenue,
    COUNT(t.last_visited) AS count_last_visited,
    COUNT(r.rating) AS count_ratings,
    COUNT(r.reviews) AS count_reviews
FROM info AS i
    INNER JOIN Finance AS f 
    ON i.product_id= f.product_id
    INNER JOIN traffic AS t 
    ON i.product_id = t.product_id
    INNER JOIN reviews as r
    ON i.product_id = r.product_id;

 * postgresql://postgres:***@localhost/retail_revenue
1 rows affected.


[(3179, 3117, 3120, 3120, 3120, 3120, 2928, 3120, 3120)]

### Analysis &  Recommendation
The database contains 3,179 products in total and only one coloumn *last_visited* is missing more than five percent or 251 of its values. It is recommended to 

1. Check with Webmaster if this is an error in data reporting and update data coloumn if true. 

2. If there is no error than it is recommended that inventory of these 251 items is reduced because of the lack  consumer interest.

## 4. Nike vs Adidas pricing
Our product portfolio, 3,179 products in total, consists of products from Adidas and Nike. 

How do the price points of Nike and Adidas products differ? 

Answering this question can help us build a picture of the company's stock range and customer market. I will run a query to produce a distribution of the *listing_price* and the count for each price, grouped by brand.

In [4]:
%%sql
-- Select the brand, listing_price as an integer, and a count of all products in finance 
-- Join brands to finance on product_id
-- Filter for products with a listing_price more than zero
-- Aggregate results by brand and listing_price, and sort the results by listing_price in descending order

SELECT (b.brand) AS brand,
    CAST((f.listing_price) AS INTEGER),
    COUNT(*)
FROM finance as f
    INNER JOIN brands as b
    ON f.product_id= b.product_id
WHERE f.listing_price > 0
GROUP BY b.brand,f.listing_price
ORDER BY f.listing_price DESC;

 * postgresql://postgres:***@localhost/retail_revenue
77 rows affected.


[('Adidas', 300, 2),
 ('Adidas', 280, 4),
 ('Adidas', 240, 5),
 ('Adidas', 230, 8),
 ('Adidas', 220, 11),
 ('Adidas', 200, 8),
 ('Nike', 200, 1),
 ('Adidas', 190, 7),
 ('Nike', 190, 2),
 ('Adidas', 180, 34),
 ('Nike', 180, 4),
 ('Adidas', 170, 27),
 ('Nike', 170, 14),
 ('Adidas', 160, 28),
 ('Nike', 160, 31),
 ('Adidas', 150, 41),
 ('Nike', 150, 6),
 ('Adidas', 140, 36),
 ('Nike', 140, 12),
 ('Adidas', 130, 96),
 ('Nike', 130, 12),
 ('Adidas', 120, 115),
 ('Nike', 120, 16),
 ('Adidas', 110, 91),
 ('Nike', 110, 17),
 ('Adidas', 100, 72),
 ('Nike', 100, 14),
 ('Adidas', 96, 2),
 ('Nike', 95, 1),
 ('Adidas', 90, 89),
 ('Nike', 90, 13),
 ('Adidas', 86, 7),
 ('Adidas', 85, 1),
 ('Nike', 85, 5),
 ('Adidas', 80, 322),
 ('Nike', 80, 16),
 ('Nike', 79, 1),
 ('Adidas', 76, 149),
 ('Adidas', 75, 1),
 ('Nike', 75, 7),
 ('Adidas', 70, 87),
 ('Nike', 70, 4),
 ('Adidas', 66, 102),
 ('Nike', 65, 1),
 ('Adidas', 63, 1),
 ('Adidas', 60, 211),
 ('Nike', 60, 2),
 ('Adidas', 56, 174),
 ('Adidas', 55, 2),
 ('Adidas', 53, 43),
 ('Adidas', 50, 183),
 ('Nike', 50, 5),
 ('Adidas', 48, 42),
 ('Nike', 48, 1),
 ('Adidas', 46, 163),
 ('Adidas', 45, 1),
 ('Nike', 45, 3),
 ('Adidas', 43, 51),
 ('Adidas', 40, 81),
 ('Nike', 40, 1),
 ('Adidas', 38, 24),
 ('Adidas', 36, 25),
 ('Adidas', 33, 24),
 ('Adidas', 30, 37),
 ('Nike', 30, 2),
 ('Adidas', 28, 38),
 ('Adidas', 27, 18),
 ('Adidas', 25, 28),
 ('Adidas', 23, 1),
 ('Adidas', 20, 8),
 ('Adidas', 18, 4),
 ('Adidas', 16, 4),
 ('Adidas', 15, 27),
 ('Adidas', 13, 27),
 ('Adidas', 12, 1),
 ('Adidas', 10, 11),
 ('Adidas', 9, 1)]

## 5. Labeling price ranges

It turns out there are 77 unique prices for the products in our database, which makes the output of the last query quite difficult to analyze.

So, I build on the previous query by assigning labels to different price brackets, grouping by brand and label. I also include the total revenue for each price bracket and brand. This creates an overview of the revenue generated by Nike and Adidas in each price bracket


In [5]:
%%sql

-- Select the brand, a count of all products in the finance table, and total revenue
-- Create four labels for products based on their price range, aliasing as price_category
-- Join brands to finance on product_id and filter out products missing a value for brand
-- Group results by brand and price_category, sort by total_revenue

SELECT (b.brand) AS brand,
    COUNT(f.*),
    SUM(f.revenue) AS total_revenue,
    CASE WHEN f.listing_price < 42 THEN 'Budget'
         WHEN f.listing_price >= 42 AND f.listing_price < 74 THEN 'Average'
         WHEN f.listing_price >= 74 AND f.listing_price < 129 THEN 'Expensive'
         WHEN f.listing_price > 129 THEN 'Elite'
     END AS price_category       
FROM finance as f
    INNER JOIN brands as b
    ON f.product_id= b.product_id
WHERE b.brand IS NOT NULL
GROUP BY b.brand, price_category
ORDER BY price_category, total_revenue;

 * postgresql://postgres:***@localhost/retail_revenue
8 rows affected.


[('Nike', 16, 6623.5, 'Average'),
 ('Adidas', 1060, 3233661.0, 'Average'),
 ('Nike', 357, 595341.0, 'Budget'),
 ('Adidas', 359, 651661.44, 'Budget'),
 ('Nike', 82, 128475.586, 'Elite'),
 ('Adidas', 307, 3014316.5, 'Elite'),
 ('Nike', 90, 71843.16, 'Expensive'),
 ('Adidas', 849, 4626979.5, 'Expensive')]

### Analysis & Recommendations
<p>Interestingly, grouping products by brand and price range allows us to see that Adidas items generate more total revenue regardless of price category! Specifically, <code>"Elite"</code> Adidas products priced \$129 or more typically generate the highest revenue, so the company can potentially increase revenue by shifting their stock to have a larger proportion of these products!</p>

## 6. Average discount by brand
<p>Note we have been looking at <code>listing_price</code> so far. The <code>listing_price</code> may not be the price that the product is ultimately sold for. To understand <code>revenue</code> better, let's take a look at the <code>discount</code>, which is the percent reduction in the <code>listing_price</code> when the product is actually sold. We would like to know whether there is a difference in the amount of <code>discount</code> offered between brands, as this could be influencing <code>revenue</code>.</p>

In [6]:
%%sql
    
-- Select brand and average_discount as a percentage
-- Join brands to finance on product_id
-- Aggregate by brand
-- Filter for products without missing values for brand

SELECT (b.brand) AS brand,
    AVG(f.discount) * 100 AS average_discount
FROM brands as b
    INNER JOIN finance as f
    ON b.product_id = f.product_id
WHERE b.brand IS NOT NULL
GROUP BY b.brand
ORDER BY average_discount;

 * postgresql://postgres:***@localhost/retail_revenue
2 rows affected.


[('Nike', 0.0), ('Adidas', 33.45242746362408)]

### Analysis & Recommendations
<p>Strangely, no <code>discount</code> is offered on Nike products! In comparison, not only do Adidas products generate the most revenue, but these products are also heavily discounted! </p>
<p>To improve revenue further, the company could try to reduce the amount of discount offered on Adidas products, and monitor sales volume to see if it remains stable. Alternatively, it could try offering a small discount on Nike products. This would reduce average revenue for these products, but may increase revenue overall if there is an increase in the volume of Nike products sold. </p>

## 7. Correlation between revenue and reviews

<p>Now I will explore whether relationships exist between the revenue and customer reviews. To do this I will check the strength and direction of a correlation between <code>revenue</code> and <code>reviews</code>. </p>

In [7]:
%%sql

-- Calculate the correlation between reviews and revenue as review_revenue_corr
-- Join the reviews and finance tables on product_id

SELECT corr(r.reviews,f.revenue) as review_revenue_corr
FROM reviews as r
 INNER JOIN finance as f
    ON r.product_id = f.product_id;

 * postgresql://postgres:***@localhost/retail_revenue
1 rows affected.


[(0.6518512280081782,)]

### Analysis & Recommendations
<p>Interestingly, there is a strong positive correlation between <code>revenue</code> and <code>reviews</code>. This means, potentially, if we can get more reviews on the company's website, it may increase the sales of items with a higher number of reviews. </p>

## 8. Ratings and reviews by product description length
<p>Perhaps the length of a product's <code>description</code> might influence a product's <code>rating</code> and <code>reviews</code> &mdash; if so, the company can produce content guidelines for listing products on their website and test if this influences <code>revenue</code>. Let's check this out!</p>

In [8]:
%%sql

-- Calculate description_length
-- Convert rating to a numeric data type and calculate average_rating
-- Join info to reviews on product_id and group the results by description_length
-- Filter for products without missing values for description, and sort results by description_length

SELECT
TRUNC(LENGTH(i.description), -2) AS description_length,
ROUND (AVG (r.rating :: numeric), 2) AS average_rating
FROM info as i
    INNER JOIN reviews as r
    ON i.product_id= r.product_id
WHERE i.description IS NOT NULL
GROUP BY description_length
ORDER BY description_length;

 * postgresql://postgres:***@localhost/retail_revenue
7 rows affected.


[(Decimal('0'), Decimal('1.87')),
 (Decimal('100'), Decimal('3.21')),
 (Decimal('200'), Decimal('3.27')),
 (Decimal('300'), Decimal('3.29')),
 (Decimal('400'), Decimal('3.32')),
 (Decimal('500'), Decimal('3.12')),
 (Decimal('600'), Decimal('3.65'))]

### Analysis & Recommendations
<p>There doesn't appear to be a clear relationship between the length of a product's <code>description</code> and its <code>rating</code>.</p>
<p>However, we do know that a positive correlation exists between <code>reviews</code> and <code>revenue</code>. So we could encourage customers to leave more reviews about their purchases by, for example, offering customers discount codes for taking the time to submit a review for their purchases.<p>

## 9. Reviews by month and brand

<p>Let's take a look at the volume of <code>reviews</code> by month to see if there are any trends or gaps we can look to exploit.</p>

In [9]:
%%sql

-- Select brand, month from last_visited, and a count of all products in reviews aliased as num_reviews
-- Join traffic with reviews and brands on product_id
-- Group by brand and month, filtering out missing values for brand and month
-- Order the results by brand and month

SELECT
    (b.brand) AS brand,
    EXTRACT(month FROM t.last_visited) AS month,
    COUNT(r.*) AS num_reviews
FROM traffic AS t
   INNER JOIN reviews AS r
   ON t.product_id= r.product_id
   INNER JOIN brands AS b
   ON t.product_id = b.product_id
WHERE b.brand IS NOT NULL AND EXTRACT(month FROM t.last_visited) IS NOT NULL
GROUP BY brand, month
ORDER BY brand, month ;

 * postgresql://postgres:***@localhost/retail_revenue
24 rows affected.


[('Adidas', Decimal('1'), 253),
 ('Adidas', Decimal('2'), 272),
 ('Adidas', Decimal('3'), 269),
 ('Adidas', Decimal('4'), 180),
 ('Adidas', Decimal('5'), 172),
 ('Adidas', Decimal('6'), 159),
 ('Adidas', Decimal('7'), 170),
 ('Adidas', Decimal('8'), 189),
 ('Adidas', Decimal('9'), 181),
 ('Adidas', Decimal('10'), 192),
 ('Adidas', Decimal('11'), 150),
 ('Adidas', Decimal('12'), 190),
 ('Nike', Decimal('1'), 52),
 ('Nike', Decimal('2'), 52),
 ('Nike', Decimal('3'), 55),
 ('Nike', Decimal('4'), 42),
 ('Nike', Decimal('5'), 41),
 ('Nike', Decimal('6'), 43),
 ('Nike', Decimal('7'), 37),
 ('Nike', Decimal('8'), 29),
 ('Nike', Decimal('9'), 28),
 ('Nike', Decimal('10'), 47),
 ('Nike', Decimal('11'), 38),
 ('Nike', Decimal('12'), 35)]

### Analysis & Recommendations
<p>Looks like product reviews are highest in the first quarter of the calendar year, possibly an impact of Christmas shopping. So there is scope to run experiments aiming to increase the volume of reviews in the other nine months! For example, discount codes can be offered in exchange for reviews from Q2-Q4. </p>

## 10. Footwear product performance

<p>So far, I have been primarily analyzing Adidas vs Nike products. Now, let's switch our attention to the type of products being sold. As there are no labels for product type, we will create a Common Table Expression (CTE) that filters <code>description</code> for keywords, then use the results to find out how much of the company's stock consists of footwear products and the median <code>revenue</code> generated by these items.</p>

In [10]:
%%sql

-- Create the footwear CTE, containing description and revenue
-- Filter footwear for products with a description containing %shoe%, %trainer, or %foot%
-- Also filter for products that are not missing values for description
-- Calculate the number of products and median revenue for footwear products

WITH footwear AS
( 
 SELECT 
    i.description,
    f.revenue  
    FROM info AS i
    INNER JOIN finance AS f
        ON i.product_id = f. product_id
    WHERE i.description ILIKE  '%shoe%' 
        OR i.description ILIKE '%trainer%' 
        OR  i.description ILIKE '%foot%' 
        AND i.description IS NOT NULL
)

SELECT
    Count(*) AS num_footwear_products,
    Percentile_disc (0.5) WITHIN GROUP (ORDER BY revenue) AS median_footwear_revenue
FROM footwear;

 * postgresql://postgres:***@localhost/retail_revenue
1 rows affected.


[(2700, 3118.36)]

## 11. Clothing product performance
<p>Recall from the first task that we found there are 3,117 products without missing values for <code>description</code>. Of those, 2,700 are footwear products, which accounts for around 85% of the company's stock. They also generate a median revenue of over $3000 dollars!</p>
<p>This is interesting, but we have no point of reference for whether footwear's <code>median_revenue</code> is good or bad compared to other products. So, for our final task, let's examine how this differs to clothing products. We will re-use <code>footwear</code>, adding a filter afterward to count the number of products and <code>median_revenue</code> of products that are not in <code>footwear</code>.</p>

In [11]:
%%sql

-- Copy the footwear CTE from the previous task
-- Calculate the number of products in info and median revenue from finance
-- Inner join info with finance on product_id
-- Filter the selection for products with a description not in footwear
WITH footwear AS
(
    SELECT i.description, f.revenue
    FROM info AS i
    INNER JOIN finance AS f 
        ON i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%'
        AND i.description IS NOT NULL
)

SELECT COUNT(i.*) AS num_clothing_products, 
    percentile_disc(0.5) WITHIN GROUP (ORDER BY f.revenue) AS median_clothing_revenue
FROM info AS i
INNER JOIN finance AS f on i.product_id = f.product_id
WHERE i.description NOT IN (SELECT description FROM footwear);

 * postgresql://postgres:***@localhost/retail_revenue
1 rows affected.


[(417, 503.82)]

### Analysis & Recommendations
<p> Code executed in the two preceeding blocks reveals that Footwear is the cash cow for this online retailer. It has a larger number of SKUs and higher median revenue. Therefore, all the actions suggested above, for example: stock optimisation, discounts etc. are more likely to have a noticeable effect when applied to footwear products.</p>